lib/alp_workloads/kvm_workload_utils.pm

```diff
@@ -0,0 +1,201 @@
+  collect_kvm_container_setup_logs
+  setup_kvm_container_from_scratch
+  clean_and_resetup_kvm_container
+);
+
+# Container engine: can be podman(default), docker(not tested)
+our $ct_engine = "podman";
+
+my $DEFAULT_KVM_IMAGE = "registry.opensuse.org/suse/alp/workloads/tumbleweed_containerfiles/suse/alp/workloads/kvm:latest";
+our $kvm_image = $DEFAULT_KVM_IMAGE;
+
+sub set_ct_engine {
+    $ct_engine = shift;
+
+    die "Container engine can only be set to podman or docker!" if ($ct_engine ne "podman" && $ct_engine ne "docker");
+}
+
+sub set_kvm_image {
+    my $_image = shift;
+
+    # Allow customization in case of special test needs
+    if ($_image ne $DEFAULT_KVM_IMAGE) {
+        $kvm_image = $_image;
+        record_info("The kvm container image has been changed from default!", "Now test is using kvm image: $kvm_image.");
+    }
+}
+
+sub pull_kvm_container {
+    # Ensure large enough time is given to download
+    # TODO: uncomment once automation is done
+    # assert_script_run("$ct_engine pull $kvm_image", 1800/get_var('TIMEOUT_SCALE', 1));
+    record_info("Kvm container image is successfully downloaded.", "Image: $kvm_image.");
+
+}
+
+sub config_host_and_kvm_container {
+    # Copy contents from kvm container to host to set it up
+    assert_script_run("clear; $ct_engine container runlabel install $kvm_image");
+    save_screenshot;
+
+    # Replace /etc/kvm-container.conf IMAGE value with $kvm_image if non-default
+    if ($kvm_image ne $DEFAULT_KVM_IMAGE) {
+        assert_script_run("cp /etc/kvm-container.conf /etc/kvm-container.conf.org");
+        assert_script_run(q@sed -i 's#^\s*IMAGE=.*$#IMAGE=@ . $kvm_image . q@#g' /etc/kvm-container.conf@);
+        assert_script_run("grep ^IMAGE= /etc/kvm-container.conf");
+        save_screenshot;
+        record_info("/etc/kvm-container.conf default IMAGE is changed!", "New value is IMAGE=$kvm_image.");
+    }
+
+    # Share /dev/sshserial from host to kvm container for let openqa root-ssh console works
+    assert_script_run("cp /usr/local/bin/kvm-container-manage.sh /usr/local/bin/kvm-container-manage.sh.org");
+    assert_script_run(q@sed -i 's#podman create#podman create --volume /dev/sshserial:/dev/sshserial #g' /usr/local/bin/kvm-container-manage.sh@);
+    script_run("diff /usr/local/bin/kvm-container-manage.sh /usr/local/bin/kvm-container-manage.sh.org");
+    save_screenshot;
+    record_info("kvm-container-manage.sh has been changed to share sol console needed device with kvm container.");
+    assert_script_run("kvm-container-manage.sh create");
+    save_screenshot;
+    record_info("KVM container is created!");
+}
+
+sub cleanup_host_and_kvm_container {
+    assert_script_run("if $ct_engine ps | grep libvirtd; then kvm-container-manage.sh stop;fi");
+    assert_script_run("if $ct_engine ps --all | grep libvirtd; then kvm-container-manage.sh rm;fi");
+    assert_script_run("if which kvm-container-manage.sh;then kvm-container-manage.sh uninstall;fi");
```

Comment:
I do not understand this
```
if which kvm-container-manage.sh
```

Reply:
if kvm-container-manage.sh exists. 
This cleanup function can be run on whatever environment, kvm container deployed, half-cleaned, or not exist. This is to ensure assert_script_run returns correct value.